### ISSUE TITLE
Can't achieve decent throughput on localhost

### ISSUE THREAD

**Issue by jean-airoldie** [sctplab/usrsctp#390]
I can't manage to get decent localhost performance. In my application code I get about `~50Mb/sec` (tested on 2 machines). Moreover, I seems like most of the time I'm waiting for the sendv & recvv to unblock.

My benchmark is simply a socket bound to localhost trying to sendv messages in a tight loop to another thread calling recvv, also in a tight loop. Note that I'm using the library as a cffi, however I don't think this is caused by the overhead of the cffi call.

My experimental setup:
* Ubuntu 18.04 (thought I also tried on centos7)
* Clang 8.0.0
* Benchmarking using `iperf3` I get `~500Mb/s` SCTP throughput and `~10Gb/s` UDP throughput.
* Compiled usrsctp with cmake in `Release` mode with `-O3` c flag & statically linked.
* UDP encapsulation at 9899 (but I also tried with the SCTP transport)
* Using 4Mb sendspace, recvspace, `RCV_BUF` and `SND_BUF` size.
* Using one-to-one sockets in non-blocking mode (but I also tried blocking mode).
* Tried varying `sendv` message size
* Tried varying application recv buffer size (the one provided to the `recvv` call)
* Tried varying the congestion algo just in case.

My `sendv` calls looks like this:
```c
usrsctp_sendv(
  sock,
  data,
  data_len,
  NULL,
  0,
  sendv_info, // a zero'ed sctp_sendv_spa with all flags cleared
  sizeof(sctp_sendv_spa),
  0
);
```
And my `recvv` calls:
```c
usrsctp_recvv(
  sock,
  data,
  data_len,  // 4Mb
  addr,
  sizeof(sctp_storage),
  info,
  sizeof(sctp_rcvinfo),
  0,
  0
);
```
I tried profiling with perf but I didn't find anything that looked especially abnormal, probably due to the multi-threaded nature of the lib.

Surely there must be something crucial I forgot.

**Comment by lgrahl** [comment]
Same as #245 (even though that went slightly off-topic)

**Comment by ivanhxq** [comment]
I also meet the same issue and urgently needed to be resolved!

**Comment by msvoelker** [comment]
I used the tsctp program to reproduce the issue.

    git clone https://github.com/sctplab/usrsctp
    mkdir usrsctp/build
    cd usrsctp/build
    ccmake .. (disable sctp_debug)
    cmake --build .
    sudo programs/tsctp -L 127.0.0.1

In a second terminal:

    sudo programs/tsctp -T 60 127.0.0.1

This sends SCTP packets for 60 seconds and then prints the throughput. With Ubuntu 19.10 on a small sized computer I got about 62 Mbit/s. The tool mpstat showed me that at least one of the four cores of the AMD GX-412TC CPU was fully utilized.

I switched to another computer with Ubuntu 19.10 that has a faster CPU (Intel Xeon E5-2650) and repeated the test. On that computer I got a throughput of about 445 Mbit/s.

It looks like the CPU determines the throughput.
@jean-airoldie Can you confirm that. Do you remember which hardware you used for your test?

**Comment by tuexen** [comment]
@msvoelker What happens if you use UDP encapsulation. If you are running native SCTP all stacks  get all packets. So if you run a client and server on the same host, both nodes handle all packets, also the ones not intended to be received by them but by the peer.

**Comment by tuexen** [comment]
@msvoelker I used the FreeBSD kernel stack on the same slow computer (PC-Engines box), and I get about 720Mbit/s, when using a message size of 1024 bytes. When using a user message size of 16336 bytes (which corresponds to full sized frames), I get about 1.4 GBit/sec. The communication is CPU bound. A non-debug FreeBSD kernel is used.

**Comment by msvoelker** [comment]
> What happens if you use UDP encapsulation? If you are running native SCTP all stacks get all packets.

That's a good point. Back on the slow computer, I test it with calling

    sudo programs/tsctp -T 60 -E 9898 -U 9899 127.0.0.1

in the sender terminal. I got a throughput of about 68 Mbit/s.

Changing the message size instead brings a bigger performance boost. For this test, I used a message size of 1452 bytes (leads to a packet size of 1500 bytes including SCTP and IPv4 overhead). The call

    sudo programs/tsctp -T 60 -l 1452 127.0.0.1

showed a throughput of about 85 Mbit/s. I wasn't able to convince usrsctp that the MTU is higher than 1500 bytes. Anyhow, It should made clear that the message size has an impact here.

@jean-airoldie Can you confirm that?

**Comment by jean-airoldie** [comment]
> Do you remember which hardware you used for your test?

Pretty old and slow CPUs for both tests.

> Can you confirm that?

I'm not using SCTP anymore since I moved to QUIC and I don't have my original setup anymore.

**Comment by msvoelker** [comment]
OK. Just for the record before the issue gets closed, I summarize my findings.

By default, checksum calculation over loopback is disabled. However, tsctp enables it. I removed that line 

    usrsctp_sysctl_set_sctp_no_csum_on_loopback(0);

in tsctp, which gave me another performance boost.

Since the communication over loopback is highly CPU bound, the result depends on the hardware. A larger message size saves computation power. Ideally, the size of one message completely fills an IP packet.

Here are my results using standard usrsctp (except for MTU configuration) with standard tsctp (except for the disabled checksum calculation for loopback) on a slow computer with a AMD GX-412TC CPU and on a faster computer with an Intel Xeon E5-2650 CPU using this tsctp call:

    sudo programs/tsctp -T 60 -l msg_len 127.0.0.1

MTU [B] | msg_len [B] -> throughput AMD GX-412TC | throughput Intel Xeon E5-2650
1500 | 20 -> 10 Mbit/s | 43 Mbit/s
1500 | 1000 -> 80 Mbit/s | 560 Mbit/s
1500 | 1452 -> 116 Mbit/s | 972 Mbit/s
9000 | 1000 -> 208 Mbit/s | 1330 Mbit/s
9000 | 8952 -> 490 Mbit/s | 3135 Mbit/s
16384 | 1000 -> 270 Mbit/s | 1582 Mbit/s
16384 | 16336 -> 690 Mbit/s | 4300 Mbit/s
65536 | 1000 -> 386 Mbit/s | 1730 Mbit/s
65536 | 65392 -> 1214 Mbit/s | 5917 Mbit/s
65536 | 65393 -> 3 Mbit/s | 3 Mbit/s

Notes:
* To be able to send packets larger than 16384, I increased MAXLEN_MBUF_CHAIN in usrsctplib/user_socket.c from 32 to 128.
* For msg_len above 65392 bytes, something seems to dramatically decrease the throughput.

**Comment by tuexen** [comment]
I bumped MAXLEN_MBUF_CHAIN to 128 in https://github.com/sctplab/usrsctp/commit/4b5798faf9a5cbce1dfed817b87bedc935fb65cb.

The numbers provided by @msvoelker seem to make sense and not indicated a bug for me. So I'm closing this now. If you feel there is a performance problem (which could be), please open another issue of reopen this with a setup to reproduce the issue.
